sql/tests: TestRandomSyntaxGeneration failed [unhandled kind: 0]

sql/tests.TestRandomSyntaxGeneration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/17523456?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/17523456?buildTab=artifacts#/) on master @ [becbd0fcdfa2e37a6ff23b33af70f2f91eca0790](https://github.com/cockroachdb/cockroach/commits/becbd0fcdfa2e37a6ff23b33af70f2f91eca0790):

Random syntax error:

```
    rsg_test.go:962: Crash detected: server panic: pq: internal error: building declarative schema change targets for DROP SCHEMA: unhandled kind: 0
```
Query:

```
        DROP SCHEMA IF EXISTS ident , ident . PUBLIC , WORK , POLYGON . ident RESTRICT;
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #133510 sql/tests: TestRandomSyntaxGeneration failed [C-test-failure O-robot P-2 T-sql-foundations branch-release-23.2]
- #132558 sql/tests: TestRandomSyntaxGeneration failed [C-test-failure O-robot P-2 T-db-server branch-release-24.1.6-rc]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxGeneration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-43820

sql/tests.TestRandomSyntaxGeneration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/17750378?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/17750378?buildTab=artifacts#/) on master @ [1759878c80e15663634321a45efdcb52490b3e46](https://github.com/cockroachdb/cockroach/commits/1759878c80e15663634321a45efdcb52490b3e46):


```
*   | 	pkg/sql/sem/tree/type_check.go:355
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planner).analyzeExpr
*   | 	pkg/sql/analyze_expr.go:34
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planner).planTenantSpec
*   | 	pkg/sql/tenant_spec.go:63
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planner).LookupTenantInfo
*   | 	pkg/sql/tenant_spec.go:148
*   | github.com/cockroachdb/cockroach/pkg/ccl/crosscluster/physical.alterReplicationJobHook.func1
*   | 	pkg/ccl/crosscluster/physical/alter_replication_job.go:220
*   | github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec.func1
*   | 	pkg/sql/planhook.go:194
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	pkg/util/stop/stopper.go:498
*   | runtime.goexit
*   | 	src/runtime/asm_amd64.s:1695
* Wraps: (4) ReturnType called on TypedExpr with empty typeAnnotation. Was the underlying Expr type-checked before asserting a type of TypedExpr?
* Error types: (1) *withstack.withStack (2) *assert.withAssertionFailure (3) *withstack.withStack (4) *errutil.leafError
*
panic: ReturnType called on TypedExpr with empty typeAnnotation. Was the underlying Expr type-checked before asserting a type of TypedExpr? [recovered]
	panic: ReturnType called on TypedExpr with empty typeAnnotation. Was the underlying Expr type-checked before asserting a type of TypedExpr?

goroutine 1766593 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x454aa5?, {0x896b670, 0xc013b8d350})
	pkg/util/stop/stopper.go:226 +0x65
panic({0x6976880?, 0xc01caabf20?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.typeAnnotation.assertTyped(...)
	pkg/sql/sem/tree/expr.go:148
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*Subquery).TypeCheck(0xc013b8c9f0?, {0xd188020?, 0xc013b8c9f0?}, 0xc00da0aa98?, 0x99fa37?)
	pkg/sql/sem/tree/type_check.go:1754 +0xd3
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*ArrayFlatten).TypeCheck(0xc020def980, {0x896b670?, 0xc013b8d350?}, 0x896b670?, 0xe80000c02d470f30?)
	pkg/sql/sem/tree/type_check.go:1934 +0x46
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.TypeCheck({0x896b670?, 0xc013b8d350?}, {0x896c4e0?, 0xc020def980?}, 0x70596d8?, 0xc0016876b0?)
	pkg/sql/sem/tree/type_check.go:345 +0x8a
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.TypeCheckAndRequire({0x896b670?, 0xc013b8d350?}, {0x896c4e0?, 0xc020def980?}, 0xc00da0ac00?, 0xca14aa0, {0x711e66c, 0x21})
	pkg/sql/sem/tree/type_check.go:355 +0x54
github.com/cockroachdb/cockroach/pkg/sql.(*planner).analyzeExpr(0xc027eb26b8, {0x896b670, 0xc013b8d350}, {0x896c4e0?, 0xc020def980?}, {{0x0, 0x0, 0x0}, {0x0, 0x0}}, ...)
	pkg/sql/analyze_expr.go:34 +0xaf
github.com/cockroachdb/cockroach/pkg/sql.(*planner).planTenantSpec(0xc013b8d350?, {0x896b670?, 0xc013b8d350?}, 0x1e?, {0x711e66c?, 0x4000000000000000?})
	pkg/sql/tenant_spec.go:63 +0x190
github.com/cockroachdb/cockroach/pkg/sql.(*planner).LookupTenantInfo(0xc027eb26b8, {0x896b670, 0xc013b8d350}, 0x8948de0?, {0x711e66c?, 0xc00da0ae10?})
	pkg/sql/tenant_spec.go:148 +0x27
github.com/cockroachdb/cockroach/pkg/ccl/crosscluster/physical.alterReplicationJobHook.func1({0x896b670, 0xc013b8d350}, {0xc006353450?, 0xc00da0ae90?, 0xd64e94?}, 0xc00b2dc360)
	pkg/ccl/crosscluster/physical/alter_replication_job.go:220 +0x159
github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec.func1({0x896b670, 0xc013b8d350})
	pkg/sql/planhook.go:194 +0x4a
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x70bb649?, 0x0?})
	pkg/util/stop/stopper.go:498 +0x1f0
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 1762144
	pkg/util/stop/stopper.go:488 +0x49c
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #134752 sql/tests: TestRandomSyntaxGeneration failed [building declarative schema change targets for DROP SCHEMA: unhandled kind: 0] [C-test-failure O-robot P-2 T-sql-foundations branch-release-24.2]
- #134280 sql/tests: TestRandomSyntaxGeneration failed [building declarative schema change targets for DROP SCHEMA: unhandled kind: 0] [C-test-failure O-robot P-2 T-sql-foundations branch-release-24.3]
- #133510 sql/tests: TestRandomSyntaxGeneration failed [unhandled kind: 0] [C-test-failure O-robot P-2 T-sql-foundations branch-release-23.2]
- #132558 sql/tests: TestRandomSyntaxGeneration failed [CREATE TENANT statement is not valid after parsing then formatting] [C-test-failure O-robot P-2 T-db-server branch-release-24.1.6-rc]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxGeneration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/drop_schema.go
+++ b/pkg/sql/drop_schema.go
@@ -70,6 +70,14 @@ func (p *planner) DropSchema(ctx context.Context, n *tree.DropSchema) (planNode,
 			}
 			return nil, pgerror.Newf(pgcode.InvalidSchemaName, "unknown schema %q", scName)
 		}
+		hasOwnership, err := p.HasOwnership(ctx, sc)
+		if err != nil {
+			return nil, err
+		}
+		if !hasOwnership {
+			return nil, pgerror.Newf(pgcode.InsufficientPrivilege,
+				"must be owner of schema %s", tree.Name(sc.GetName()))
+		}
 
 		if scName == catconstants.PublicSchemaName {
 			return nil, pgerror.Newf(pgcode.InvalidSchemaName, "cannot drop schema %q", scName)
@@ -79,14 +87,6 @@ func (p *planner) DropSchema(ctx context.Context, n *tree.DropSchema) (planNode,
 		case catalog.SchemaPublic, catalog.SchemaVirtual, catalog.SchemaTemporary:
 			return nil, pgerror.Newf(pgcode.InvalidSchemaName, "cannot drop schema %q", scName)
 		case catalog.SchemaUserDefined:
-			hasOwnership, err := p.HasOwnership(ctx, sc)
-			if err != nil {
-				return nil, err
-			}
-			if !hasOwnership {
-				return nil, pgerror.Newf(pgcode.InsufficientPrivilege,
-					"must be owner of schema %s", tree.Name(sc.GetName()))
-			}
 			namesBefore := len(d.objectNamesToDelete)
 			fnsBefore := len(d.functionsToDelete)
 			if err := d.collectObjectsInSchema(ctx, p, db, sc); err != nil {

--- a/pkg/sql/logictest/testdata/logic_test/drop_schema
+++ b/pkg/sql/logictest/testdata/logic_test/drop_schema
@@ -21,3 +21,22 @@ BEGIN;
 ALTER TYPE schema_123539.enum_123539 DROP VALUE 's';
 DROP SCHEMA schema_123539 CASCADE;
 COMMIT;
+
+# Check that we block dropping the public schema of the system database, as
+# well as virtual schemas.
+
+statement error must be owner of schema public
+DROP SCHEMA system.public
+
+statement error must be owner of schema pg_catalog
+DROP SCHEMA pg_catalog
+
+user testuser
+
+statement error must be owner of schema public
+DROP SCHEMA system.public
+
+statement error must be owner of schema crdb_internal
+DROP SCHEMA crdb_internal
+
+user root

--- a/pkg/sql/schemachanger/scbuild/builder_state.go
+++ b/pkg/sql/schemachanger/scbuild/builder_state.go
@@ -309,7 +309,7 @@ func (b *builderState) mustOwn(id catid.DescID) {
 	b.ensureDescriptor(id)
 	if c := b.descCache[id]; !c.hasOwnership {
 		panic(pgerror.Newf(pgcode.InsufficientPrivilege,
-			"must be owner of %s %s", c.desc.DescriptorType(), c.desc.GetName()))
+			"must be owner of %s %s", c.desc.DescriptorType(), tree.Name(c.desc.GetName())))
 	}
 }
 
@@ -989,8 +989,18 @@ func (b *builderState) checkOwnershipOrPrivilegesOnSchemaDesc(
 	case catalog.SchemaTemporary:
 		// Nothing needs to be done.
 	case catalog.SchemaPublic, catalog.SchemaVirtual:
-		panic(pgerror.Newf(pgcode.InsufficientPrivilege,
-			"%s permission denied for schema %q", p.RequiredPrivilege.DisplayName(), name))
+		if p.RequireOwnership {
+			if ok, err := b.auth.HasOwnership(b.ctx, sc); err != nil {
+				panic(err)
+			} else if !ok {
+				panic(pgerror.Newf(pgcode.InsufficientPrivilege,
+					"must be owner of schema %s", tree.Name(name.Schema())))
+			}
+		} else {
+			if err := b.auth.CheckPrivilege(b.ctx, sc, p.RequiredPrivilege); err != nil {
+				panic(err)
+			}
+		}
 	case catalog.SchemaUserDefined:
 		b.ensureDescriptor(sc.GetID())
 		if p.RequireOwnership {

